# Day 13: Transparent Origami

https://adventofcode.com/2021/day/13

## Part 1

In [1]:
import numpy as np

In [2]:
example_txt = """6,10
0,14
9,10
0,3
10,4
4,11
6,0
6,12
4,1
0,13
10,12
3,4
3,0
8,4
1,10
2,14
8,10
9,0

fold along y=7
fold along x=5"""

In [3]:
with open('input.txt') as input_file:
    input_txt = input_file.read()

In [4]:
def parse_text(txt):
    """Parse the marked dots and the folding instructions."""
    
    # Store the marked dots as a list of (x, y) tuples.
    marked_dots, instructions = txt.split('\n\n')
    marked_dots = marked_dots.split()
    marked_dots = [tuple(map(int, dot.split(','))) for dot in marked_dots]
    
    # Store the folds as a list of (coordinate, value) tuples.
    # Infer the maximum x and y values from twice the value of the first folds.
    xmax, ymax = (0, 0)
    folds = []
    instructions = instructions.strip().split('\n')
    for instruction in instructions:
        coordinate, value = instruction.split()[-1].split('=')
        value = int(value)
        if coordinate == 'x' and 2*value > xmax:
            xmax = 2*value
        if coordinate == 'y' and 2*value > ymax:
            ymax = 2*value
        folds.append((coordinate, value))
    
    # Define the sheet of transparent paper and mark the positions of the dots.
    paper = np.zeros((xmax+1, ymax+1), dtype=bool)
    for dot in marked_dots:
        paper[dot[0], dot[1]] = True
    
    return marked_dots, folds, paper

In [5]:
def follow_instructions(folds, paper):
    """Fold the paper according to the instructions."""
    for fold in folds:
        coordinate, value = fold
        xmax = len(paper) - 1
        ymax = len(paper[0]) - 1
        if coordinate == 'x':  # fold along line "x = value"
            assert not any(paper[value, :])  # check no dots on fold line
            for y in range(0, ymax+1):
                for x in range(0, value):
                    paper[x, y] = paper[x, y] or paper[xmax - x, y]
            paper = paper[:value, :]
        elif coordinate == 'y':  # fold along line "y = value"
            assert not any(paper[:, value])  # check no dots on fold line
            for x in range(0, xmax+1):
                for y in range(0, value):
                    paper[x, y] = paper[x, y] or paper[x, ymax - y]
            paper = paper[:, :value]
        print(f'After folding along {coordinate}={value}, {paper.sum()} dots are visible.')
    return paper

In [6]:
def visualize_paper(paper):
    """Display a simple visualisation,
    with '*' for marked dots
    and '.' for unmarked ones."""
    for y in range(len(paper[0])):
        line = []
        for x in range(len(paper)):
            line.append('#' if paper[x, y] else '.')
        print(''.join(line))

Check the code with the example text, then generate the answers from the input text.

In [7]:
for txt in (example_txt, input_txt):
    marked_dots, folds, paper = parse_text(txt)
    paper = follow_instructions(folds, paper)
    visualize_paper(paper)
    print()

After folding along y=7, 17 dots are visible.
After folding along x=5, 16 dots are visible.
#####
#...#
#...#
#...#
#####
.....
.....

After folding along x=655, 795 dots are visible.
After folding along y=447, 653 dots are visible.
After folding along x=327, 533 dots are visible.
After folding along y=223, 450 dots are visible.
After folding along x=163, 374 dots are visible.
After folding along y=111, 308 dots are visible.
After folding along x=81, 249 dots are visible.
After folding along y=55, 199 dots are visible.
After folding along x=40, 166 dots are visible.
After folding along y=27, 139 dots are visible.
After folding along y=13, 110 dots are visible.
After folding along y=6, 88 dots are visible.
.##..####...##.#..#.#....#..#..##....##.
#..#.#.......#.#.#..#....#..#.#..#....#.
#....###.....#.##...#....#..#.#.......#.
#....#.......#.#.#..#....#..#.#.##....#.
#..#.#....#..#.#.#..#....#..#.#..#.#..#.
.##..####..##..#..#.####..##...###..##..



## Part 1

Q. How many dots are visible after completing just the first fold instruction on your transparent paper?

A. 795

## Part 2

Q. What code do you use to activate the infrared thermal imaging camera system?

A. CEJKLUGJ